src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -0,0 +1,266 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ODataUtf8JsonWriter.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP3_1_OR_GREATER
+using System;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Globalization;
+using System.Text.Encodings.Web;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// Implement of <see cref="IJsonWriter"/> that is based on
+    /// <see cref="Utf8JsonWriter"/>.
+    /// </summary>
+    internal class ODataUtf8JsonWriter : IJsonWriter, IDisposable
+    {
+        private const int DefaultBufferSize = 16 * 1024;
+        private const float BufferFlushThreshold = 0.9f;
+        private readonly static byte[] Parentheses = new byte[] { (byte)'(', (byte)')' };
+
+        private readonly Stream outputStream;
+        private readonly Utf8JsonWriter writer;
+        private readonly int bufferSize;
+        private readonly bool isIeee754Compatible;
+        private readonly bool leaveStreamOpen;
+        private readonly Encoding encoding = Encoding.UTF8;
+        private bool disposed;
+
+        public ODataUtf8JsonWriter(Stream outputStream, bool isIeee754Compatible, int bufferSize = DefaultBufferSize, bool leaveStreamOpen = false)
+        {
+            this.outputStream = outputStream;
+            this.isIeee754Compatible = isIeee754Compatible;
+            this.bufferSize = bufferSize;
+            this.writer = new Utf8JsonWriter(
+                outputStream,
+                // we don't need to perform validation here since the higher-level
+                // writers already perform validation
+                new JsonWriterOptions { SkipValidation = true });
+            this.leaveStreamOpen = leaveStreamOpen;
+        }
+
+        public void Flush()
+        {
+            if (this.writer.BytesPending > 0)
+            {
+                this.writer.Flush();
+            }
+        }
+
+        public void FlushIfBufferThresholdReached()
+        {
+            // flush when we're close to the buffer capacity to avoid
+            // allocating bigger buffers
+            if (this.writer.BytesPending >= BufferFlushThreshold * this.bufferSize)
+            {
+                this.Flush();
+            }
+        }
+
+        public void StartPaddingFunctionScope()
+        {
+            this.Flush();
+            this.outputStream.Write(Parentheses, 0, 1);

```

Comment:
I'm not sure what I'm missing here, but would `this.outputStream.WriteByte((byte)'(');` be clearer about the intention and use (ever so slightly) less memory?

Reply:
@corranrogue9 thanks, that's exactly what I was looking for. I didn't know Stream provided a `WriteByte()` API 😄